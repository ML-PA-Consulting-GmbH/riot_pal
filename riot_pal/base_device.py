# Copyright (c) 2018 Kevin Weiss, for HAW Hamburg  <kevin.weiss@haw-hamburg.de>
#
# This file is subject to the terms and conditions of the MIT License. See the
# file LICENSE in the top level directory for more details.
# SPDX-License-Identifier:    MIT
"""Abstract Base Device for RIOT PAL
This module is the abstract device that interfaces to a driver.  It handles
the selection and initialzation of the driver.

Example:
    class ApplicationDevice(BaseDevice):
        def show_example(self):
            self._write('Send Example')
            print(self._read())

    example_use_case = ApplicationDevice('serial', port='my/port/name')
    example_use_case.show_example()
"""
import logging
try:
    from . import driver_manager
except SystemError:
    import driver_manager


class BaseDevice:
    """Instance for devices to connect and utilize drivers.

    Args:
        *args: Variable length argument list.
        **kwargs: Arbitrary keyword arguments.
    """

    def __init__(self, *args, **kwargs):
        self._driver = driver_manager.driver_from_config(*args, **kwargs)

    def close(self):
        """Closes the device connection."""
        self._driver.close()

    def _read(self):
        """Reads data from the driver.

        Returns:
            str: string of data if success, ERR string if failed
        """
        return self._driver.read()

    def _write(self, data):
        """Writes data to the driver.

        Args:
            data(str): Variable length argument list.
        """
        return self._driver.write(data)

    def is_connected_to_board(self):
        """Dummy - confirm if a connection is for target board."""
        logging.warning("Check if board is connected: dummy should be"
                        " implmeneted in subclasses")
        raise NotImplementedError()

    @classmethod
    def from_autodetect(cls, *args, **dev_config):
        """Connects to a range of possible configurations."""
        configs = driver_manager.available_configs(*args, **dev_config)
        logging.debug("Configs: %r", configs)
        for config in configs:
            for retry in range(0, 2):
                logging.debug("Autodetect attempt: %d", retry)
                conn = cls(**config)
                try:
                    if conn.is_connected_to_board():
                        return conn
                except Exception as err:
                    logging.debug("Cannot connect: %r", err)
                conn.close()

        raise ValueError("Could not locate board, check if board is"
                         "connected or is_connected_to_board is correct")

    @classmethod
    def copy_driver(cls, device):
        """Copies the driver instance so many devices can use one driver."""
        logging.debug("Cloning Driver: %r", device._driver)
        return cls(dev_type='driver', driver=device._driver)


def main():
    """Tests basic usage of the class

    Used for unit testing, information should be confirm with DEBUG info.
    """
    logging.getLogger().setLevel(logging.DEBUG)
    BaseDevice()


if __name__ == "__main__":
    main()
