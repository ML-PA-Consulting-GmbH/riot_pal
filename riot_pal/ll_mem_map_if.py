# Copyright (c) 2018 Kevin Weiss, for HAW Hamburg  <kevin.weiss@haw-hamburg.de>
#
# This file is subject to the terms and conditions of the MIT License. See the
# file LICENSE in the top level directory for more details.
# SPDX-License-Identifier:    MIT
"""Low Level Memory Map Interface for RIOT PAL
This module handles offset and sizes dictated a memory map.
"""
import logging
import csv
try:
    from .ll_shell import LLShell
except SystemError:
    from ll_shell import LLShell


class LLMemMapIf(LLShell):
    """Interface between ll_shell and mem_map."""

    def __init__(self, path, *args, **kwargs):
        self.cmd_list = LLMemMapIf.import_mm_from_csv(path)
        super().__init__(*args, **kwargs)

    @staticmethod
    def import_mm_from_csv(path):
        """Imports a memory map csv file."""
        cmd_list = {}
        with open(path) as csvfile:
            rows = list(csv.reader(csvfile, quotechar="'"))
            for row in range(1, len(rows)):
                cmd = dict(zip(rows[0], rows[row]))
                cmd_list[rows[row][rows[0].index('name')]] = cmd
                logging.debug("Imported command: %r", cmd)
        return cmd_list

    def read_struct(self, cmd_name):
        """Reads a set of registers defined by the memory map."""
        response = []
        for cmd, val in self.cmd_list.items():
            # Supress unused variable warning
            val = val
            if cmd.startswith(cmd_name):
                response.append(self.read_reg(cmd))
        return response

    def read_reg(self, cmd_name, offset=0, size=None):
        """Read a register defined by the memory map.

        Keyword arguments:
        offset -- the number of elements to offset in an array
        size   -- the number of elements to read in an array
        """
        cmd = self.cmd_list[cmd_name]
        response = None
        if size is None:
            size = cmd['total_size']
        if cmd['is_bitfield'] is 'True':
            response = self.read_bits(cmd['offset'],
                                      cmd['bit_offset'],
                                      cmd['bits'])
        elif cmd['size'] != cmd['total_size']:
            offset = int(cmd['offset']) + (offset * int(cmd['size']))
            if size is None:
                size = cmd['total_size']
            else:
                size = size * int(cmd['size'])
            response = self.read_bytes(offset, size)
        else:
            response = self.read_bytes(cmd['offset'], cmd['total_size'])
        response['msg'] = 'cmd={} response={}'.format(cmd_name,
                                                      response['msg'])
        return response

    def write_reg(self, cmd_name, data, offset=0):
        """Writes a register defined by the memory map.

        Keyword arguments:
        offset -- the number of elements to offset in an array
        """
        cmd = self.cmd_list[cmd_name]
        response = None
        if cmd['is_bitfield'] is 'True':
            response = self.write_bits(cmd['offset'],
                                       cmd['bit_offset'],
                                       cmd['bits'], data)
        elif cmd['size'] != cmd['total_size']:
            offset = int(cmd['offset']) + (offset * int(cmd['size']))
            response = self.write_bytes(offset, data)
        else:
            response = self.write_bytes(cmd['offset'], data, int(cmd['size']))
        response['msg'] = 'cmd={} response={}'.format(cmd_name,
                                                      response['msg'])
        return response


def main():
    """Tests all functions supported functions."""
    from pprint import pprint
    logging.getLogger().setLevel(logging.DEBUG)

    mmif = LLMemMapIf("mem_map.csv", 'serial', '/dev/ttyUSB0')
    print('==================================================================')
    try:
        pprint(mmif.read_struct('sys'))
        pprint(mmif.read_reg('user_reg.64', 0, 10))
        pprint(mmif.write_reg('user_reg.64', [6, 6, 6], 2))
        pprint(mmif.write_reg('user_reg.64', [6, 6, 6], 2))
        pprint(mmif.read_reg('user_reg.64'))
    except KeyError:
        print("WARNING: Unrecognized mem_map")
    pprint(mmif.cmd_list)
    for cmd, val in mmif.cmd_list.items():
        # Supress unused variable warning
        val = val
        pprint(cmd)
        pprint(mmif.read_reg(cmd))


if __name__ == "__main__":
    main()
