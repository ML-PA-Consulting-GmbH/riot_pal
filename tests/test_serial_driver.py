# Copyright (c) 2018 Kevin Weiss, for HAW Hamburg  <kevin.weiss@haw-hamburg.de>
#
# This file is subject to the terms and conditions of the MIT License. See the
# file LICENSE in the top level directory for more details.
# SPDX-License-Identifier:    MIT
"""Tests Serial Driver implmentation in RIOT PAL."""
from pprint import pformat
from serial import SerialException
from riot_pal.serial_driver import SerialDriver

WORKING_PORT = '/dev/ttyACM0'
WORKING_BAUD = 115200
FAILING_PORT = 'ERROR'
FAILING_BAUD = 19200


def _open_write_read_close(regtest, *args, **kwrgs):
    regtest.write('Opening with args:\n')
    regtest.write(pformat(args) + '\n')
    regtest.write(pformat(kwrgs) + '\n')
    try:
        ser_drvr = SerialDriver(*args, **kwrgs)
        ser_drvr.write('')
        regtest.write(ser_drvr.read())
        ser_drvr.close()
        regtest.write('SUCCESS\n')
    except (SerialException, TypeError) as exc:
        regtest.write(str(exc) + '\n')


def test_connect_working(regtest):
    """Test working serial connections."""
    test_port = WORKING_PORT
    test_baud = WORKING_BAUD
    _open_write_read_close(regtest)
    _open_write_read_close(regtest, test_port)
    _open_write_read_close(regtest,
                           port=test_port)
    _open_write_read_close(regtest,
                           test_port,
                           test_baud)
    _open_write_read_close(regtest,
                           test_port,
                           baudrate=test_baud)
    _open_write_read_close(regtest,
                           baudrate=test_baud,
                           port=test_port,
                           timeout=2)


def test_connect_failing(regtest):
    """Test failing serial connections."""
    test_port = FAILING_PORT
    test_baud = FAILING_BAUD
    _open_write_read_close(regtest, test_port)
    _open_write_read_close(regtest,
                           port=test_port)
    _open_write_read_close(regtest,
                           test_port,
                           test_baud)
    _open_write_read_close(regtest,
                           test_port,
                           baudrate=test_baud)
    _open_write_read_close(regtest,
                           baudrate=test_baud,
                           port=test_port,
                           timeout=2)
