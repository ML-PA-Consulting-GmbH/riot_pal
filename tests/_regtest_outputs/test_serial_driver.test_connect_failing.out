Opening with args:
('ERROR',)
{}
[Errno 2] could not open port ERROR: [Errno 2] No such file or directory: 'ERROR'
Opening with args:
()
{'port': 'ERROR'}
serial_for_url() missing 1 required positional argument: 'url'
Opening with args:
('ERROR', 19200)
{}
[Errno 2] could not open port ERROR: [Errno 2] No such file or directory: 'ERROR'
Opening with args:
('ERROR',)
{'baudrate': 19200}
[Errno 2] could not open port ERROR: [Errno 2] No such file or directory: 'ERROR'
Opening with args:
()
{'baudrate': 19200, 'port': 'ERROR', 'timeout': 2}
serial_for_url() missing 1 required positional argument: 'url'
